### PR TITLE
Heappe paradigm

### PR DESCRIPTION
**Pull request by kozusznik**

### CODE REVIEW

File: src/main/java/cz/it4i/parallel/Constants.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package cz.it4i.parallel;
+
+interface Constants {
+	long TEMPLATE_ID = 6;
+	int TIMEOUT = 3600;
+	long CLUSTER_NODE_TYPE = 7l;
+	String PROJECT_ID = "open-12-20";
+	String CONFIG_FILE_NAME = "configuration.properties";
+	String JOB_NAME = "ImageJ-Server-Heappee";
```
Comment (PetrBainar) [inline src/main/java/cz/it4i/parallel/Constants.java]: It'd be good to distinguish what's applicable for the entire package (e.g. configuration.properties?) and what's specific for a given paradigm (e.g. job name "ImageJ-Server-Heappee" [sic!, there is a typo btw]).

Consequently, consider moving heappe-paradigm-specific constants to the paradigm class.

File: src/main/java/cz/it4i/parallel/HeappeParadigm.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+@Plugin(type = ParallelizationParadigm.class)
+public class HeappeParadigm extends SimpleOstravaParadigm {
+
+	private static final int TIMEOUT_BETWEEN_JOB_STATE_QUERY = 1000;
+
+	@Parameter
+	private int port;
+
+	public static final Logger log = LoggerFactory.getLogger(cz.it4i.parallel.HeappeParadigm.class);
```
Comment (PetrBainar) [inline src/main/java/cz/it4i/parallel/HeappeParadigm.java]: Consider making this variable private, it won't raise any issues in the IDE anymore.

File: src/main/java/cz/it4i/parallel/HeappeParadigm.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+	private int numberOfHosts;
```
Comment (PetrBainar) [inline src/main/java/cz/it4i/parallel/HeappeParadigm.java]: Just thinking whether we shouldn't treat this instance variable in the same way as we treat the port number. If you concur, consider adding the @Parameter decoration.

File: src/main/java/cz/it4i/parallel/HeappeParadigm.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+		this.port = port;
+	}
+
+	public void setNumberOfNodes(int number) {
+		this.numberOfHosts = number;
+	}
+
+	// -- SimpleOstravaParadigm methods --
+
+	// -- Closeable methods --
```
Comment (PetrBainar) [inline src/main/java/cz/it4i/parallel/HeappeParadigm.java]: Closeable methods shall be defined further down in the file, see line 106.

File: src/main/java/cz/it4i/parallel/HeappeParadigm.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+				TunnelToNode tunnel;
+				tunnels.add(tunnel = haasClient.openTunnel(jobId, node, 0, port));
+				return new HeappeWorker(tunnel.getLocalHost(), tunnel.getLocalPort());
+			}).forEach(worker -> workerPool.addWorker(worker));
+		} else {
+			log.error("Job ID: %d not running. It is in state %s.", jobId, state.toString());
+		}
+	}
+
+	private JobState logGetState(long jobId) {
```
Comment (PetrBainar) [inline src/main/java/cz/it4i/parallel/HeappeParadigm.java]: Consider moving this method to the helper methods section.

File: src/main/java/cz/it4i/parallel/HeappeParadigm.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+				t.close();
+			} catch (IOException e) {
+				log.error(e.getMessage(), e);
+			}
+		});
+	}
+
+	@Override
+	protected void initWorkerPool() {
+		if (log.isDebugEnabled()) {
```
Comment (PetrBainar) [inline src/main/java/cz/it4i/parallel/HeappeParadigm.java]: According to this [link](https://stackoverflow.com/questions/963492/in-log4j-does-checking-isdebugenabled-before-logging-improve-performance), the `isDebugEnable()` guard statement is redundant in this case.

There are multiple occurrences of the issue in this method.

File: src/main/java/cz/it4i/parallel/HeappeParadigm.java (hunk 5 above)
Comment (PetrBainar) [inline src/main/java/cz/it4i/parallel/HeappeParadigm.java]: This method has a single purpose and that is obtaining the job state. Logging for debugging purposes does not have to be included in the method name in my opinion.

File: src/main/java/cz/it4i/parallel/HeappeParadigm.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+			log.debug("initWorkerPool");
+		}
+		haasClient = new HaaSClient(SettingsProvider.getSettings(Constants.PROJECT_ID, Constants.CONFIG_FILE_NAME));
+		if (log.isDebugEnabled()) {
+			log.debug("createJob");
+		}
+		long jobId = haasClient.createJob(
```
Comment (PetrBainar) [inline src/main/java/cz/it4i/parallel/HeappeParadigm.java:58]: Consider making this local variable final. The same suggestions applies to multiple occurrences in this method.

File: src/main/java/org/scijava/parallel/Node.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
+package org.scijava.parallel;
+
+import java.util.LinkedList;
+import java.util.List;
+
+public abstract class Node<T> {
```
Comment (PetrBainar) [inline src/main/java/org/scijava/parallel/Node.java:6]: Can this file be moved to a dedicated branch? It is not used at the moment at all. It was intended to be used for defining dependencies between individual tasks, but this feature has not been implemented yet.

Also, this is probably my fault, sorry for this.

File: src/test/java/test/learnathon/Demonstration.java [hunk 9]
```diff
@@ -68,7 +68,7 @@ public void run() {
 		hosts.add("localhost:10001");
 		hosts.add("localhost:10002");
 		hosts.add("localhost:10003");
-		hosts.add("localhost:10004");
+		hosts.add("localhost:10004");
```
Comment (PetrBainar) [inline src/test/java/test/learnathon/Demonstration.java]: Consider discarding this change.

**Review by PetrBainar — COMMENTED** [review]
Checking WIP review - still 4 files pending.

File: src/main/java/org/scijava/parallel/Node.java (hunk 8 above)
Comment (kozusznik) [reply src/main/java/org/scijava/parallel/Node.java:6]: I suggest not to complicate it and leave it on this branch.

File: src/main/java/cz/it4i/parallel/HeappeParadigm.java (hunk 5 above)
Comment (kozusznik) [reply src/main/java/cz/it4i/parallel/HeappeParadigm.java]: OK

Comment (kozusznik) [reply src/main/java/cz/it4i/parallel/HeappeParadigm.java]: OK

File: src/main/java/cz/it4i/parallel/HeappeParadigm.java (hunk 4 above)
Comment (kozusznik) [reply src/main/java/cz/it4i/parallel/HeappeParadigm.java]: OK

File: src/main/java/cz/it4i/parallel/HeappeParadigm.java (hunk 3 above)
Comment (kozusznik) [reply src/main/java/cz/it4i/parallel/HeappeParadigm.java]: I don't know. Probably yes, I didn't solve this in the issue. I will change it.

File: src/main/java/cz/it4i/parallel/HeappeParadigm.java (hunk 2 above)
Comment (kozusznik) [reply src/main/java/cz/it4i/parallel/HeappeParadigm.java]: OK, I give up and I will do logger private in the future.

File: src/main/java/cz/it4i/parallel/Constants.java (hunk 1 above)
Comment (kozusznik) [reply src/main/java/cz/it4i/parallel/Constants.java]: OK

File: src/main/java/cz/it4i/parallel/HeappeParadigm.java (hunk 7 above)
Comment (kozusznik) [reply src/main/java/cz/it4i/parallel/HeappeParadigm.java:58]: According to the discussion https://stackoverflow.com/questions/316352/why-would-one-mark-local-variables-and-method-parameters-as-final-in-java we can discuss it and use it as a  global rule. However, I don't think as an issue of this merge request.
Let's discuss code convention rules and create rules for automatic changes in Eclipse IDE.

Comment (PetrBainar) [reply src/main/java/cz/it4i/parallel/HeappeParadigm.java:58]: Agree, created an issue (https://github.com/PetrBainar/scijava-parallel/issues/9) lest we forget it.

File: src/main/java/org/scijava/parallel/Node.java (hunk 8 above)
Comment (PetrBainar) [reply src/main/java/org/scijava/parallel/Node.java:6]: OK

**Comment by kozusznik** [comment]
I see only 2 files not outdated

**Comment by PetrBainar** [comment]
Review finished. Feel free to hit the button.

**Pull request merged**